// user-onboarding/lib/component/step/carousel/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var preact_1 = require("preact");
var svg_1 = require("./svg");
var StepCarousel = /** @class */ (function (_super) {
    __extends(StepCarousel, _super);
    function StepCarousel() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            index: 0,
        };
        _this.handlePre = function () {
            var index = _this.state.index - 1;
            _this.setState({ index: index });
        };
        _this.handleNext = function () {
            var index = _this.state.index + 1;
            _this.setState({ index: index });
        };
        _this.handleFinish = function () {
            _this.props.onNext();
        };
        return _this;
    }
    Object.defineProperty(StepCarousel.prototype, "currentCarouselChild", {
        get: function () {
            var item = this.props.data.children[this.state.index];
            // 暂时只支持 img
            return preact_1.h("img", { src: item });
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StepCarousel.prototype, "isTheFirst", {
        get: function () {
            return this.state.index === 0;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(StepCarousel.prototype, "isTheLast", {
        get: function () {
            return this.state.index === this.props.data.children.length - 1;
        },
        enumerable: true,
        configurable: true
    });
    StepCarousel.prototype.render = function () {
        var _a = this.props.data, topOffset = _a.topOffset, leftOffset = _a.leftOffset, width = _a.width;
        var wrapperStyle = {
            position: 'fixed',
            top: '0',
            right: '0',
            bottom: '0',
            left: '0',
        };
        var contentStyle = {
            position: 'absolute',
            top: '50%',
            left: '50%',
            transform: 'translate(-50%, -50%)',
            padding: '16px',
            minWidth: '200px',
            width: width,
            marginTop: topOffset && topOffset + "px",
            marginLeft: leftOffset && leftOffset + "px",
            border: '1px solid #ECECEC',
            borderRadius: '4px',
            boxShadow: '0px 4px 20px rgba(0, 0, 0, 0.39)',
            background: '#fff',
        };
        var preStyle = {
            position: 'absolute',
            top: '50%',
            left: '-50px',
            width: '24px',
            height: '24px',
            transform: 'translateY(-50%)',
            cursor: 'pointer',
            fill: '#ccc',
        };
        var nextStyle = {
            position: 'absolute',
            top: '50%',
            right: '-50px',
            width: '24px',
            height: '24px',
            transform: 'translateY(-50%)',
            cursor: 'pointer',
            fill: '#ccc',
        };
        return (preact_1.h("div", { style: wrapperStyle },
            preact_1.h("div", { style: contentStyle },
                !this.isTheFirst && preact_1.h(svg_1.Left, { style: preStyle, onClick: this.handlePre }),
                this.currentCarouselChild,
                this.isTheLast ? (preact_1.h("span", { style: nextStyle, onClick: this.handleFinish }, "\u5B8C\u6210")) : (preact_1.h(svg_1.Right, { style: nextStyle, onClick: this.handleNext })))));
    };
    return StepCarousel;
}(preact_1.Component));
exports.StepCarousel = StepCarousel;
