// user-onboarding/lib/component/mask/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
var preact_1 = require("preact");
var Mask = /** @class */ (function (_super) {
    __extends(Mask, _super);
    function Mask() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Mask.prototype.componentDidMount = function () {
        document.body.style.overflow = 'hidden';
        document.body.style.height = '100vh';
        document.body.style.width = '100vw';
    };
    Mask.prototype.componentWillUnmount = function () {
        document.body.style.overflow = 'auto';
        document.body.style.height = 'auto';
        document.body.style.width = 'auto';
    };
    Mask.prototype.getOffset = function () {
        var _a = this.props, targetDom = _a.targetDom, area = _a.area;
        if (targetDom) {
            return targetDom.getBoundingClientRect();
        }
        if (area) {
            return area;
        }
    };
    Mask.prototype.render = function () {
        var offset = this.getOffset();
        var wrapperStyle = {
            position: 'fixed',
            left: 0,
            top: 0,
            height: '100vh',
            width: '100vw',
            zIndex: 1000,
            pointerEvents: 'none',
        };
        var maskCommonStyle = {
            position: 'absolute',
            background: 'rgba(0,0,0,.5)',
            pointerEvents: 'initial',
        };
        return (preact_1.h("div", { style: wrapperStyle }, offset ? ([
            // top
            preact_1.h("div", { key: '1', style: __assign({}, maskCommonStyle, { top: 0, left: 0, right: 0, height: offset.top }) }),
            // right
            preact_1.h("div", { key: '2', style: __assign({}, maskCommonStyle, { top: offset.top, left: offset.right, right: 0, height: offset.bottom - offset.top }) }),
            // bottom
            preact_1.h("div", { key: '3', style: __assign({}, maskCommonStyle, { top: offset.bottom, left: 0, right: 0, bottom: 0 }) }),
            // left
            preact_1.h("div", { key: '4', style: __assign({}, maskCommonStyle, { top: offset.top, left: 0, width: offset.left, height: offset.bottom - offset.top }) }),
        ]) : (preact_1.h("div", { style: __assign({}, maskCommonStyle, { top: 0, left: 0, right: 0, bottom: 0 }) }))));
    };
    return Mask;
}(preact_1.Component));
exports.Mask = Mask;
