// user-onboarding/lib/component/step/notice/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var preact_1 = require("preact");
var StepNotice = /** @class */ (function (_super) {
    __extends(StepNotice, _super);
    function StepNotice() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    StepNotice.prototype.render = function () {
        var _a = this.props.data, content = _a.content, leftOffset = _a.leftOffset, topOffset = _a.topOffset, width = _a.width;
        var wrapperStyle = {
            position: 'fixed',
            top: '0',
            right: '0',
            bottom: '0',
            left: '0',
        };
        var contentStyle = {
            position: 'absolute',
            top: '50%',
            left: '50%',
            transform: 'translate(-50%, -50%)',
            padding: '16px',
            marginTop: topOffset && topOffset + "px",
            marginLeft: leftOffset && leftOffset + "px",
            minWidth: '200px',
            width: width,
            border: '1px solid #ECECEC',
            borderRadius: '4px',
            boxShadow: '0px 4px 20px rgba(0, 0, 0, 0.39)',
            background: '#fff',
        };
        var footerStyle = {
            marginTop: '16px',
            textAlign: 'right',
        };
        return (preact_1.h("div", { style: wrapperStyle },
            preact_1.h("div", { style: contentStyle },
                preact_1.h("div", { dangerouslySetInnerHTML: { __html: content } }),
                this.props.progressChild && (preact_1.h("footer", { style: footerStyle }, this.props.progressChild)))));
    };
    return StepNotice;
}(preact_1.Component));
exports.StepNotice = StepNotice;
