// user-onboarding/lib/app.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var preact_1 = require("preact");
var step_1 = require("./component/step");
var hoc_1 = require("./component/step/hoc");
var debug_1 = require("./debug");
var theme_1 = require("./component/provider/theme");
var input_checker_1 = require("./component/step/input-checker");
var stepComponentMap = {
    base: null,
    notice: step_1.StepNotice,
    focus: step_1.StepFocus,
    carousel: step_1.StepCarousel,
    inputChecker: input_checker_1.StepInputChecker,
};
function getStepComponent(type) {
    var Comp = stepComponentMap[type];
    if (!Comp)
        return;
    return hoc_1.withAllStepHelper(Comp);
}
var CANCEL_FLAG = -1;
var App = /** @class */ (function (_super) {
    __extends(App, _super);
    function App() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.state = {
            stepIndex: 0,
        };
        _this.handleNext = function () {
            _this.gotoNext();
        };
        _this.handleCancel = function () {
            _this.cancel();
        };
        return _this;
    }
    App.prototype.gotoNext = function () {
        var stepIndex = this.state.stepIndex + 1;
        this.setState({ stepIndex: stepIndex });
        this.props.onNextStep && this.props.onNextStep(stepIndex);
        debug_1.normalDebug.log('gotoNext: stepIndex=%s', stepIndex);
        if (stepIndex >= this.props.sceneData.steps.length) {
            this.props.onFinish && this.props.onFinish();
            debug_1.normalDebug.log('scene is finish');
        }
    };
    App.prototype.cancel = function () {
        var stepIndex = CANCEL_FLAG;
        this.setState({ stepIndex: stepIndex });
    };
    App.prototype.render = function () {
        var sceneData = this.props.sceneData;
        var stepIndex = this.state.stepIndex;
        var currentStep = sceneData.steps[stepIndex];
        if (!currentStep) {
            debug_1.normalDebug.info('currentStep is null');
            return null;
        }
        debug_1.normalDebug.info('currentStep: type=%s, index=%s', currentStep.type, stepIndex);
        var StepComp = getStepComponent(currentStep.type);
        return (preact_1.h(theme_1.ThemeContext.Provider, { value: sceneData.theme }, StepComp && (preact_1.h(StepComp
        // 赋予不同的 key，保证每个 Step 都是新的（生命周期里有副作用）
        , { 
            // 赋予不同的 key，保证每个 Step 都是新的（生命周期里有副作用）
            key: stepIndex, data: currentStep, stepIndex: stepIndex, stepTotal: sceneData.steps.length, onNext: this.handleNext, onCancel: this.handleCancel }))));
    };
    return App;
}(preact_1.Component));
exports.App = App;
