// user-onboarding/lib/debug.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var debug_1 = __importDefault(require("debug"));
var scopeDebug = debug_1.default('onboarding');
function makeBizDebugBase(name) {
    return /** @class */ (function () {
        function Base() {
            var _this = this;
            this.default = scopeDebug.extend(name);
            this.log = this.default.extend('log');
            this.info = this.default.extend('info');
            this.warn = this.default.extend('warn');
            this._error = this.default.extend('error');
            this.error = function (formatter) {
                var args = [];
                for (var _i = 1; _i < arguments.length; _i++) {
                    args[_i - 1] = arguments[_i];
                }
                console.error.apply(console, [formatter].concat(args));
                _this._error.apply(_this, [formatter].concat(args));
            };
        }
        return Base;
    }());
}
var NormalDebug = /** @class */ (function (_super) {
    __extends(NormalDebug, _super);
    function NormalDebug() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return NormalDebug;
}(makeBizDebugBase('normal')));
/** 通用 debug */
exports.normalDebug = new NormalDebug();
