// user-onboarding/lib/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
Object.defineProperty(exports, "__esModule", { value: true });
var preact_1 = require("preact");
var app_1 = require("./app");
var pub_1 = require("./pub");
var debug_1 = require("./debug");
__export(require("./service"));
function insertMountRoot(id) {
    if (id === void 0) { id = 'user-onboarding-root-' + new Date().valueOf(); }
    var root = document.createElement('div');
    root.id = id;
    document.body.appendChild(root);
    return root;
}
function createApp(config) {
    var _this = this;
    var mount = insertMountRoot();
    return function (name) {
        var pub = new pub_1.PubCenter();
        var renderRoot;
        var api = {
            /** 启动 */
            start: function () {
                debug_1.normalDebug.log('Onboarding start: name=%s, mount=%s', name, mount.id);
                (function () { return __awaiter(_this, void 0, void 0, function () {
                    var sceneData;
                    return __generator(this, function (_a) {
                        switch (_a.label) {
                            case 0: return [4 /*yield*/, config.service.getScene(name)];
                            case 1:
                                sceneData = _a.sent();
                                debug_1.normalDebug.info('Onboarding data ready: name=%s, stepCnt=%s', name, sceneData.steps.length);
                                renderRoot = preact_1.render(preact_1.h(app_1.App, { sceneData: sceneData, onNextStep: function () {
                                        var args = [];
                                        for (var _i = 0; _i < arguments.length; _i++) {
                                            args[_i] = arguments[_i];
                                        }
                                        return pub.publish.apply(pub, ['nextStep'].concat(args));
                                    }, onFinish: function () { return pub.publish('finish'); } }), mount);
                                return [2 /*return*/];
                        }
                    });
                }); })();
                return api;
            },
            /** 停止 */
            stop: function () {
                // How to destroy root Preact node
                // @see https://github.com/developit/preact/issues/1151
                preact_1.render(null, mount, renderRoot);
                debug_1.normalDebug.log('Onboarding stop');
            },
            /** 监听事件 */
            on: function (topic, handler) {
                pub.subscribe(topic, handler);
                return api;
            },
            /** 取消监听事件 */
            off: function (topic, handler) {
                pub.unSubscribe(topic, handler);
                return api;
            },
        };
        return api;
    };
}
exports.createApp = createApp;
